File: Level 2/05. Tries/208. Implement Trie.cpp
class Node
{
    public:
    char data;
    unordered_map<char,Node*> children;
    bool terminal;
    Node(char data)
    {
        this->data = data;
        terminal = false;
    }
};
class Trie {
public:
    Node* root;
    Trie() { root = new Node('\0'); }
    
    void insert(string s) 
    {
        Node* temp = root;
        for(int i=0;i < s.size();i++)
        {
            if(temp->children.find(s[i]) != temp->children.end())
                temp = temp->children[s[i]];
            else 
            {
                Node* newNode = new Node(s[i]);
                temp->children[s[i]] = newNode;
                temp = newNode;
            }
        }
        temp->terminal = true;
    }
    
    bool search(string s) 
    {
        Node* temp = root;
        for(int i=0;i < s.size();i++)
        {
            if(temp->children.find(s[i]) == temp->children.end())
                return false;    
            temp = temp->children[s[i]];
        }
        return temp->terminal;
    }
    
    bool startsWith(string s) 
    {
        Node* temp = root;
        for(int i=0;i<s.size();i++)
        {
            if(temp->children.find(s[i]) == temp->children.end())
                return false;    
            temp = temp->children[s[i]];
        }
        return true;
    }
};

File: Level 2/05. Tries/208. Implement Trie.java
class Trie {
    public static class Node {
        private Node[] children = new Node[26];
        private boolean isEnd = false;

        public boolean contains(char ch) {
            return (children[ch - 'a'] != null);
        }

        public Node get(char ch) {
            return children[ch - 'a'];
        }

        public void set(char ch) {
            children[ch - 'a'] = new Node();
        }

        public boolean getEnd() {
            return isEnd;
        }

        public void setEnd() {
            isEnd = true;
        }
    }

    Node root;

    public Trie() {
        root = new Node();
    }

    public void insert(String word) {
        Node curr = root;
        for (int i = 0; i < word.length(); i++) {
            char ch = word.charAt(i);

            if (curr.contains(ch) == false)
                curr.set(ch);

            curr = curr.get(ch);
        }

        curr.setEnd();
    }

    public boolean search(String word) {
        Node curr = root;
        for (int i = 0; i < word.length(); i++) {
            char ch = word.charAt(i);

            if (curr.contains(ch) == false)
                return false;

            curr = curr.get(ch);
        }

        return curr.getEnd();
    }

    public boolean startsWith(String word) {
        Node curr = root;
        for (int i = 0; i < word.length(); i++) {
            char ch = word.charAt(i);

            if (curr.contains(ch) == false)
                return false;

            curr = curr.get(ch);
        }

        return true;
    }
}